### PR TITLE
[KYUUBI #1989] Decouple curator from other modules

### PR DESCRIPTION
**Pull request by hddong**
### _Why are the changes needed?_

Decouple curator from other modules.

Mainly changes:
1. `ZookeeperClientProvider` -> `zookeeper.ClientProvider`
2. `zkClient` apis -> `DiscoveryClient`
3. `ZooKeeperClientProvider.withZkClient` -> `DiscoveryClientProvider.withDiscoveryClient`

### _How was this patch tested?_
- [X] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [ ] Add screenshots for manual tests if appropriate

- [ ] [Run test](https://kyuubi.apache.org/docs/latest/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by pan3793** [comment]
The Hudi flaky test has been resolved, please rebase on master branch

**Comment by hddong** [comment]
@pan3793 : thanks for your remind, troubled by the failure of the test :)

**Comment by ulysses-you** [comment]
@hddong  please resolve the conflicts since the master branch changed...

**Comment by hddong** [comment]
@ulysses-you Thanks, resolved.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/DiscoveryClient.scala [hunk 1]
```diff
@@ -0,0 +1,198 @@
...
+
+import org.apache.kyuubi.KyuubiSQLException
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.ha.client.ServiceNodeInfo
+import org.apache.kyuubi.ha.client.zookeeper.ClientProvider.buildZookeeperClient
+import org.apache.kyuubi.ha.client.zookeeper.DiscoveryClient.parseInstanceHostPort
+import org.apache.kyuubi.ha.client.zookeeper.ServiceDiscoveryClient.createServiceNode
+
+class DiscoveryClient(conf: KyuubiConf) extends Logging {
```
Comment (pan3793) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/DiscoveryClient.scala:40]: In design of #1991, `DiscoveryClient` should be an interface

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/DiscoveryClient.scala [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+import org.apache.zookeeper.CreateMode
+
+import org.apache.kyuubi.KyuubiSQLException
+import org.apache.kyuubi.config.KyuubiConf
+import org.apache.kyuubi.ha.client.KyuubiDiscoveryClient
+import org.apache.kyuubi.ha.client.ServiceNodeInfo
+import org.apache.kyuubi.ha.client.zookeeper.ClientProvider.buildZookeeperClient
+import org.apache.kyuubi.ha.client.zookeeper.ServiceDiscoveryClient.createServiceNode
+
+class DiscoveryClient(conf: KyuubiConf) extends KyuubiDiscoveryClient {
```
Comment (pan3793) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/DiscoveryClient.scala:39]: Can we name the trait `DiscoveryClient` and the derived class `ZookeeperDiscoveryClient`/`EtcdDiscoveryClient` as the proposal documented?

Comment (hddong) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/DiscoveryClient.scala]: Sure, had rename.

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ServiceDiscoveryClient.scala [hunk 3]
```diff
@@ -17,40 +17,13 @@
...
-import org.apache.kyuubi.ha.HighAvailabilityConf.HA_ZK_PUBLISH_CONFIGS
+import org.apache.kyuubi.ha.client.DiscoveryClient
 import org.apache.kyuubi.ha.client.ServiceDiscovery
-import org.apache.kyuubi.ha.client.ZooKeeperClientProvider.{buildZookeeperClient, getGracefulStopThreadDelay}
-import org.apache.kyuubi.ha.client.zookeeper.ServiceDiscoveryClient.connectionChecker
-import org.apache.kyuubi.ha.client.zookeeper.ServiceDiscoveryClient.createServiceNode
-import org.apache.kyuubi.util.KyuubiHadoopUtils
-import org.apache.kyuubi.util.ThreadUtils
 
 class ServiceDiscoveryClient(serviceDiscovery: ServiceDiscovery) extends Logging {
```
Comment (pan3793) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ServiceDiscoveryClient.scala:28]: I think this class can be merged into `ServiceDiscovery` now

Comment (hddong) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/zookeeper/ServiceDiscoveryClient.scala]: > I think this class can be merged into `ServiceDiscovery` now

Sounds great, done.

**Review by pan3793 — APPROVED** [review]

**Comment by pan3793** [comment]
@yaooqinn this PR touched zk related test classes, would you like 1.5.0 include this one? if yes, i will do #2040 based on this one

**Comment by yaooqinn** [comment]
> @yaooqinn this PR touched zk related test classes, would you like 1.5.0 include this one? if yes, i will do #2040 based on this one

will you give me your reason for 1.5？

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/DiscoveryClient.scala [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+package org.apache.kyuubi.ha.client
+
+import java.util.concurrent.TimeUnit
+
+import com.google.common.annotations.VisibleForTesting
+
+import org.apache.kyuubi.Logging
+import org.apache.kyuubi.config.KyuubiConf
+
+trait DiscoveryClient extends Logging {
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/DiscoveryClient.scala:27]: The interface class should have a clear class, method, parameter, return comments

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/DiscoveryClient.scala [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.kyuubi.ha.client
+
+import java.util.concurrent.TimeUnit
+
+import com.google.common.annotations.VisibleForTesting
```
Comment (yaooqinn) [inline kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/DiscoveryClient.scala:22]: do we have to use guava here?

**Comment by Narcasserun** [comment]
What is the PR related issues

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/DiscoveryClient.scala (hunk 5 above)
Comment (hddong) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/DiscoveryClient.scala]: Removed

File: kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/DiscoveryClient.scala (hunk 4 above)
Comment (hddong) [reply kyuubi-ha/src/main/scala/org/apache/kyuubi/ha/client/DiscoveryClient.scala:28]: Done

**Comment by hddong** [comment]
@pan3793 @yaooqinn : All comments have addressed, any more review?

**Review by yaooqinn — APPROVED** [review]

**Comment by yaooqinn** [comment]
can you fix the test failures @hddong ?

**Comment by hddong** [comment]
@yaooqinn : Sure.

**Comment by hddong** [comment]
@yaooqinn All test passed.

**Review by yaooqinn — APPROVED** [review]
LGTM

**Comment by yaooqinn** [comment]
thanks, merged to master

**Pull request closed without merging**